flowmachine/flowmachine/features/subscriber/label_event_score.py

```diff
@@ -0,0 +1,192 @@
+        and day of week score on the y axis, where all values are in the range [-1, 1]
+    required :
+        Specifies a label which every subscriber must possess independently of
+        the score.  This is used in cases where, for instance, we require
+        that all subscribers must have an evening/home location.
+    """
+
+    def __init__(
+        self,
+        *,
+        scores: Union[EventScore, _JoinedHartiganCluster],
+        labels: Dict[str, Dict[str, Any]] = {
+            "evening": {
+                "type": "MultiPolygon",
+                "coordinates": [
+                    [[[0.000_001, 0.5], [0.000_001, 1], [1, 1], [1, 0.5]]],
+                    [[[0.000_001, -1], [0.000_001, -0.5], [1, -0.5], [1, -1]]],
+                ],
+            },
+            "day": {
+                "type": "Polygon",
+                "coordinates": [[[-1, -0.5], [-1, 0.5], [0, 0.5], [0, -0.5]]],
+            },
+        },
+        required: Union[str, None] = None,
+    ):
+
+        self.scores = scores
+        if not isinstance(scores, Query):
+            raise TypeError(
+                "Scores must be of type Query, e.g. EventScores, Table, CustomQuery"
+            )
+
+        labels = LabelEventScore._make_bounds_dict(labels)
+        LabelEventScore.bounds_dict_has_overlaps(labels)
+        self.labels = labels
+
+        self.label_names = list(labels.keys())
+        self.required = required
+        if "unknown" in self.label_names:
+            raise ValueError(
+                "'unknown' is a reserved label name, please use another name"
+            )
+        else:
+            self.label_names = ["unknown"] + self.label_names
+
+        super().__init__()
+
+    @staticmethod
+    def _make_bounds_dict(
+        geojson_labels: Dict[str, Dict[str, Any]]
+    ) -> Dict[str, BaseGeometry]:
+        """
+        Takes a dictionary of labels and bounds expressed as lists of geojson shapes
+        and returns a dictionary of labels and bounds expressed as Shapely polygons.
+
+        Parameters
+        ----------
+        geojson_labels : dict
+            String -> geojson mappings
+        Returns
+        -------
+        dict
+            Dict of labels mapped to lists of shapely polygons
+
+        """
+        bounds_dict = {}
+        for label, geom in geojson_labels.items():
+            try:
+                bounds_dict[label] = shape(geom)
+            except (AttributeError, IndexError, ValueError) as e:
+                raise ValueError(f"Geometry for {label} is not valid: {e}")
+        return bounds_dict
+
+    def _make_query(self):
+
+        scores_cols = self.scores.column_names
+        scores = f"({self.scores.get_query()}) AS scores"
+
+        sql = f"""
+        SELECT COALESCE(score_bounds.label, 'unknown') as label, scores.* FROM {scores}
+        LEFT JOIN {LabelEventScore._get_sql_bound(self.labels)}
+        ON st_contains(score_bounds.geom, st_point(scores.score_hour, scores.score_dow))
```

Comment:
Nitpick: I think in other places we are using capitalised versions of PostGIS commants. Since we're also uppercasing SQL commands, maybe best to stick to the versions used in the official PostGIS docs (e.g. `ST_Contains`, `ST_Point`)?